### PR TITLE
Replace mutex by atomics

### PR DESCRIPTION
**Pull request by F001**
Hi Deno team,

Thank you for creating such an amazing project. It's my pleasure to participate in.

This PR is basically trying to remove `Mutex`es from `IsolateState`.

1. Use `AtomicBool` in `DenoPermissions`, so we don't have to use `Mutex` around `DenoPermissions`.

2. Use `AtomicUsize` in `Metrics`, so we don't have to use `Mutex` around `Metrics`.
  Please note that `AtomicU64` type is still unstable, I guess `AtomicUsize` is suitable for this situation.

3. Move `mpsc::Sender<(i32, Buf)>` outside of `IsolateState`.
    * In general, we always clone a new `Sender` and move it to new thread.  `Mutex<Sender<_>>` is unidiomatic in Rust.
    * And also, I don't feel the `Option` is useful here.

4. Refine the construction of `Isolate` and `IsolateState` instances.
  The command line arguments actually only useful to `IsolateState`. It makes no sense to accept them in `Isolate::new` function. The new process is that: a) create a `IsolateState` from command line options  b) create a `Isolate` from `IsolateState`.

Hopefully this change can make Deno slightly faster, but I don't have concrete profile numbers.

And another question, I can't build debug version on my machine, only release build succeeded.
So `./tools/test.py` command fails to run on my machine. I'm using Win10's Linux subsystem.

### CODE REVIEW

**Comment by kevinkassimo** [comment]
Hi @F001 , thanks for the PR and Rust code improvements!
Not quite sure why your debug build fails. However, you can still test the release version of deno (since you got it working) by specifying the directory where deno lives in, in your case `./tools/test.py target/release/`

**Comment by F001** [comment]
@kevinkassimo Thanks for your suggestion. Now the test failures are fixed.

File: src/isolate.rs [hunk 1]
```diff
@@ -133,44 +140,25 @@ fn empty() -> libdeno::deno_buf {
...
-    flags: flags::DenoFlags,
-    argv_rest: Vec<String>,
-    dispatch: Dispatch,
-  ) -> Self {
+  pub fn new(state: Arc<IsolateState>, dispatch: Dispatch) -> Self {
     DENO_INIT.call_once(|| {
       unsafe { libdeno::deno_init() };
     });
     let shared = empty(); // TODO Use shared for message passing.
+    let snapshot = unsafe { super::snapshot::deno_snapshot.clone() };
```
Comment (ry) [inline src/isolate.rs]: The snapshot should be passed as an argument to `Isolate::new()`

File: src/msg.rs [hunk 2]
```diff
@@ -2,6 +2,20 @@
 #![allow(dead_code)]
 #![cfg_attr(feature = "cargo-clippy", allow(clippy, pedantic))]
 use flatbuffers;
+use std::sync::atomic::Ordering;
+
 // GN_OUT_DIR is set either by build.rs (for the Cargo build), or by
 // build_extra/rust/run.py (for the GN+Ninja build).
 include!(concat!(env!("GN_OUT_DIR"), "/gen/msg_generated.rs"));
+
+impl<'a> From<&'a super::isolate::Metrics> for MetricsResArgs {
```
Comment (ry) [inline src/msg.rs]: Nice. I would like to use this technique more!

File: src/isolate.rs (hunk 1 above)
Comment (F001) [reply src/isolate.rs]: Fixed. Thanks!

File: src/isolate.rs [hunk 3]
```diff
@@ -438,17 +433,18 @@ mod tests {
   fn test_metrics_sync() {
     let argv = vec![String::from("./deno"), String::from("hello.js")];
     let (flags, rest_argv, _) = flags::set_flags(argv).unwrap();
-    let mut isolate =
-      Isolate::new(empty(), flags, rest_argv, metrics_dispatch_sync);
+    let state = Arc::new(IsolateState::new(flags, rest_argv));
+    let snapshot = unsafe { crate::snapshot::deno_snapshot.clone() };
```
Comment (ry) [inline src/isolate.rs]: use `empty()` here

File: src/isolate.rs [hunk 4]
```diff
@@ -398,7 +390,10 @@ mod tests {
   fn test_dispatch_sync() {
     let argv = vec![String::from("./deno"), String::from("hello.js")];
     let (flags, rest_argv, _) = flags::set_flags(argv).unwrap();
-    let mut isolate = Isolate::new(empty(), flags, rest_argv, dispatch_sync);
+
+    let state = Arc::new(IsolateState::new(flags, rest_argv));
+    let snapshot = unsafe { crate::snapshot::deno_snapshot.clone() };
```
Comment (ry) [inline src/isolate.rs]: use `empty()` here

File: src/isolate.rs [hunk 5]
```diff
@@ -461,30 +457,31 @@ mod tests {
...
   }
 
   #[test]
   fn test_metrics_async() {
     let argv = vec![String::from("./deno"), String::from("hello.js")];
     let (flags, rest_argv, _) = flags::set_flags(argv).unwrap();
-    let mut isolate =
-      Isolate::new(empty(), flags, rest_argv, metrics_dispatch_async);
+    let state = Arc::new(IsolateState::new(flags, rest_argv));
+    let snapshot = unsafe { crate::snapshot::deno_snapshot.clone() };
```
Comment (ry) [inline src/isolate.rs]: use `empty()` here (you're changing the test behavior)

**Pull request merged**